### ISSUE TITLE
[Prebuilt Extension] Using a web worker with React results in shared module error

### ISSUE THREAD

**Issue by jqrun** [jupyterlab/jupyterlab#10197]
## Description

When developing a prebuilt extension, trying to use a web worker that utilizes React results in the following error:

```
Uncaught (in promise) Error: Shared module react doesn't exist in shared scope default
    at ensureExistence (consumes:19)
    at consumes:93
    at consumes:78
    at Object.webpack/sharing/consume/default/react (consumes:122)
    at Object.__webpack_require__.m.<computed> [as webpack/sharing/consume/default/react] (consumes:130)
    at __webpack_require__ (bootstrap:19)
    at Object../lib/worker.js (lib_worker_js.cec92441c3301ea9a8a7.js:12)
    at __webpack_require__ (bootstrap:19)
    at bootstrap:34
    at Function.__webpack_require__.O (chunk loaded:23)
```

I tried to reproduce this with some basic module federation Webpack examples with no success (though I wasn't able to get to the stage of the prebuilt federation that Jupyterlab does), so I'm checking here to see if anyone knows the root cause. 

I noticed some federation dynamic remote container code here that inits the sharing scope `default` so it may be related, though I'm not too familiar yet with module federation: https://github.com/jupyterlab/jupyterlab/blob/f0790b3ffd187b7dc923156b8ae227855abecefb/jupyterlab/staging/bootstrap.js#L80

## Reproduce

You can use this minimally modified cookiecutter mimerenderer extension here: https://github.com/jqrun/jupyterlab-federated-worker-react with instructions in the readme.

## Expected behavior

Worker should be able to use React. Note that the non-example extension with web worker/React functionality has worked well in previous versions of lab -- this was caught in the middle of upgrading to v3. Though in this case, the difference is that we're using the new Webpack 5 worker integration rather than using worker-loader due to related issue https://github.com/jupyterlab/jupyterlab/issues/10033.

## Context

- Operating System and version: MacOS Big Sur 11.3 and Red Hat Enterprise Linux Server 7.9
- Browser and version: Chrome 90.0.4430.93
- JupyterLab version: 3.0.14

**Comment by jqrun** [comment]
There's also an interesting thread here related to Webpack, web workers, and shared chunks that may have some sort of relevance in debugging: https://github.com/webpack/webpack/issues/6472.

**Comment by jtpio** [comment]
Thanks @jqrun for reporting this issue.

This does indeed appear to be an issue of the module federation system used by JupyterLab 3+ when using web workers.

I looked into this a bit and opened https://github.com/jqrun/jupyterlab-federated-worker-react/pull/1 to explore a solution.

Not sure whether there should be a change to the current JupyterLab build system to be able to address this automatically. An alternative could be to provide documentation for using Web Worker in a JupyterLab extension.

**Comment by krassowski** [comment]
> An alternative could be to provide documentation for using Web Worker in a JupyterLab extension

Yup, I had lost too many hours hitting the wall here too. Importing a module in a Web Worker with Module Federation enabled is a long-standing issue in Webpack 5:
- https://github.com/webpack/webpack/issues/16633
- https://github.com/webpack/webpack/issues/16173
- https://github.com/webpack/webpack/issues/17518

and the workarounds either don't work fully (`MarkWorkerAsInitialPlugin` helps in the sense that assets are placed in the web worker context but still fail to import) or are not achievable in JupyterLab (multiple compilation as in https://github.com/webpack/webpack/issues/16633#issuecomment-1495056534 - because we can only modify the existing set of config from `jupyterlab.webpackConfig` stanza, not add a new one. Do we want to allow extensions to append/prepend custom compiler configs?

For what it is worth, I managed to workaround it by just using dynamic import from CDN in the worker (plus type import and some black magic for type safety) but this obviously is not a very good solution.

**Comment by krassowski** [comment]
> For what it is worth, I managed to workaround it by just using dynamic import from CDN in the worker (plus type import and some black magic for type safety) but this obviously is not a very good solution.

(and it only works in Chromium not in Firefox because of privacy/security guards I suppose)
